### PR TITLE
Flask Upgrades

### PR DESCRIPTION
**Pull request by nealmick**
This branch contains updates for the flask web server and adds functionality for team rosters as well as player details including injury status.   The main changes include 2 new templates `player_modal.html` and `team_modal.html`, these are linked to the main index.html using jinja template `{% include %}`  feature.    The flask app has been updated with new routes `/player-stats/<player_id>/` and  `/team-data/<team_name>/` which are used to request and form the data used in the new templates.  

This pr should be easier to review as its mostly HTML templates, with a few additions to the flask app.   I did have some trouble with the NN runner, and had to revert it to an older version due to TensorFlow mismatch with the saved model H5 legacy format weights, however i can remove those changes as well if needed.

https://github.com/user-attachments/assets/fde03caf-c76a-49f1-8d0b-b9babca4dcfe

### CODE REVIEW

**Comment by nealmick** [comment]
okay appreciate it, let me know if you want any changes

**Comment by nealmick** [comment]
You are correct nothing is cached between page reloads.  I do have a subscription attached to the key with a higher rate limit, it has 1k request/day, I use this key for other projects too.   Currently the api usage is at 2.8% for the month.   Caching would definitely help reduce the api hits and potential issues with rate limit.   When the page first loads it shouldn't make any extra requests, only when clicking on teams it triggers the team roster request, and then clicking on a player will hit the 2 player endpoints.   Adding cache is a good idea, i tend to use the python pickle library for saving and caching.   I would use the team ABV and playerID as cache keys.   The only issue with caching is stale data and handling cache invalidation.   I think the simplest solution would be like a clear cache button as you said, but if someone docent understand to clear the cache they could look at stale data and may think that's an issue.  Caching would reduce api hits and loading time for the team and player modals which would be good. 

I think caching would add more complexity for this pr, I would probably hold off and add caching later as long as we are still within api limits.

**Comment by XZoro-404** [comment]
To ensure that there is no abuse with your api key I would exclude it personally. But that is all up to you.

I agree that caching is outside of the scope of this pr.

But to delve a little more into the topic. I think the way to cache it would include the date and time that it was cached. Then when requesting team/player stats it will compare the current date with the cached date (this would make it so the users would not have to refresh the data daily). Then there could be a button that could do one of the following:
1.  Clear all of the current days cached data
2. Recursively clear the data from users current location ie. if on the Bulls team page and clicks refresh button only clears the data of the bulls team and player data
3. Pull all new data (do all api calls for all teams/players) 

With that said in terms of holding onto previous cached data we could either hold all data for the entire season or we could only hold it for x amount of days.

**Comment by nealmick** [comment]
Yea there is some potential for abuse, but its a hard rate limit im not too worried, can always rotate keys if needed.  Caching is definitely a good idea, im not sure how much it would help if we expire cache daily or even longer then it will get cache miss on each clients first requests.   Also the cache would not be shared between individual installs if its done locally.  Maybe could setup a remote server that would cache for all installs and remove rate limit completely, or rate-limit individual clients.

**Comment by XZoro-404** [comment]
I'm not sure which api subscription you have but if it is the 'PRO' version then there is not absolute hard cap by default.
 `Basic: This is a free plan offering a hard cap of 1,000 hits per month. Perfect for testing. This plan does NOT require a credit card.

Pro: $10/month - 1,000 requests per day with **$0.01 each addl request**.`
I don't mean to harp on this to much just want to point it out

how would it get a cache miss on clients first request? The request should be its own function and should be called on page load and compare against the already cached data's date. 

A remote server is an interesting idea, but I believe it would go against the nature of open source. For security reasons it would introduce an attack vector. Not only that but if we introduce a way for users to update the cache throughout the day then it would change it for everyone.

**Comment by nealmick** [comment]
No problem, I do have a custom plan with a hard limit and no extra overage, its the same as the 10$ just with hard limit.  

I think the cache mainly would help if someone loads the player or team data multiple times, however after 24 hours there could be new games or updated injury status so i think invalidation is good idea.  However this means the cache would only cover reloads of the same data within that 24 hour period.   Usually first loads are cache misses, unless the data is pre fetched or cache is updated regularly like every 24h, but that would need requests for every player every 24 hours and for each user separately.

I appreciate the input, security could be an issue but if you sanitize the data and do everything properly it should be okay, however adding remote server would add dependency on external service and I get it, thats not really opensource.  However i do think it would be nice to have a public hosted version of the app, with like default models so people can just click a link and check it out, but you know it all depends on where this project is going... and like you said in discussions there inst really a roadmap and so forth.

File: src/Predict/NN_Runner.py
Comment (nealmick) [inline src/Predict/NN_Runner.py:1]: includes update to fix legacy model format error.

**Comment by nealmick** [comment]
@kyleskom any update on getting this reviewed?

**Review by kyleskom — APPROVED** [review]
Tested and works as expected. Sorry it took way way too long to get to this. This is great and really appreciate this work!

**Pull request merged**